### PR TITLE
boards: nxp: rd_rw612_bga: enable flash driver support

### PR DESCRIPTION
**Pull request by danieldegrasse**
Enable support for the mx25u51245g flash chip present on the
RD RW612 BGA. Support has been verified with the following samples:

- samples/drivers/flash_shell
- tests/drivers/flash/common
- tests/boot/test_mcuboot (dependent on setting `CONFIG_BOOT_MAX_IMG_SECTORS` to 1024)

### CODE REVIEW

**Review by dleach02 — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

**Review by mmahadevan108 — APPROVED** [review]

File: drivers/clock_control/clock_control_mcux_syscon.c [hunk 1]
```diff
@@ -187,15 +187,64 @@ static int mcux_lpc_syscon_clock_control_get_subsys_rate(
...
 }
 
+/*
+ * Since this function is used to reclock the FlexSPI when running in
+ * XIP, it must be located in RAM when MEMC driver is enabled.
+ */
+#ifdef CONFIG_MEMC
+#define SYSCON_SET_FUNC_ATTR __ramfunc
+#else
+#define SYSCON_SET_FUNC_ATTR
```
Comment (mmahadevan108) [inline drivers/clock_control/clock_control_mcux_syscon.c:216]: This function is not adjusting the clock. It is handled inside `flexspi_clock_set_freq`. I think we don't need `mcux_lpc_syscon_clock_control_set_subsys_rate` to be in RAM.

File: drivers/clock_control/clock_control_mcux_syscon.c [hunk 2]
```diff
@@ -187,15 +187,64 @@ static int mcux_lpc_syscon_clock_control_get_subsys_rate(
...
+	mcux_lpc_syscon_clock_control_set_subsys_rate(const struct device *dev,
+			clock_control_subsys_t subsys,
+			clock_control_subsys_rate_t rate)
+{
+	uint32_t clock_name = (uintptr_t)subsys;
+	uint32_t clock_rate = (uintptr_t)rate;
+
+	switch (clock_name) {
+	case MCUX_FLEXSPI_CLK:
+#if defined(CONFIG_SOC_SERIES_RW6XX) && defined(CONFIG_MEMC)
```
Comment (mmahadevan108) [inline drivers/clock_control/clock_control_mcux_syscon.c:229]: Why we need `CONFIG_SOC_SERIES_RW6XX`, is `CONFIG_MEMC` not sufficient?

**Review by mmahadevan108 — CHANGES_REQUESTED** [review]
A few comments

File: drivers/clock_control/clock_control_mcux_syscon.c (hunk 1 above)
Comment (danieldegrasse) [reply drivers/clock_control/clock_control_mcux_syscon.c:216]: So the reason this function is in ram is that upon return from `flexspi_clock_set_freq`, the flash may not be accessible. The call sequence to change the clock looks like so:

- call `memc_mcux_flexspi_update_clock`. This function runs from RAM
- call `clock_control_set_rate`. This function will perform a flash read to get the API pointer for the `mcux_lpc_syscon_clock_control_set_subsys_rate` function
- call `flexspi_clock_set_freq`. ***After return from this function, flash access may not be possible***. If the clock rate change was significant enough, we will  need to update the DLL value of the FlexSPI before we can perform flash access again.
- call `FLEXSPI_UpdateDLLValue` and `memc_flexspi_reset`
- wait for the FlexSPI bus to be idle. After this point, we can safely access flash again.

For this reason, I think it is best to keep the `flexspi_clock_set_freq` function in RAM. It is hard to catch the bugs that result from invalid FlexSPI reads, which is why I typically err on the side of caution while changing the FlexSPI clock.

Comment (mmahadevan108) [reply drivers/clock_control/clock_control_mcux_syscon.c:216]: What about the clock control API functions functions? Should we move those to RAM as well?

Comment (danieldegrasse) [reply drivers/clock_control/clock_control_mcux_syscon.c:216]: The `clock_control_xxx` functions? Those functions are inlined, they will be declared in the scope of the file calling them (so they will be in RAM in this case). However, the implementations are not in RAM- that is why we have to reconfigure the FlexSPI before calling `clock_control_get_rate`: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/memc/memc_mcux_flexspi.c#L120

**Review by mmahadevan108 — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

**Review by mmahadevan108 — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

**Review by EmilioCBen — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

**Comment by dleach02** [comment]
@nordic-krch the PR is ready but needs your approval

**Review by mmahadevan108 — DISMISSED** [review]

**Review by nordic-krch — DISMISSED** [review]

**Review by dleach02 — APPROVED** [review]

**Review by mmahadevan108 — APPROVED** [review]

**Review by decsny — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**